trackintel/preprocessing/staypoints.py

```diff
@@ -72,27 +84,40 @@ def generate_locations(
             db = DBSCAN(eps=epsilon, min_samples=num_samples, algorithm="ball_tree", metric=distance_metric)
 
         if agg_level == "user":
-            location_id_counter = 0
-            # TODO: change into groupby
-            for user_id_this in ret_stps["user_id"].unique():
-                # Slice staypoints array by user. This is not a copy!
-                user_staypoints = ret_stps[ret_stps["user_id"] == user_id_this]
 
-                if distance_metric == "haversine":
-                    # the input is converted to list of (lat, lon) tuples in radians unit
-                    p = np.array([[radians(g.y), radians(g.x)] for g in user_staypoints.geometry])
-                else:
-                    p = np.array([[g.x, g.y] for g in user_staypoints.geometry])
-                labels = db.fit_predict(p)
+            # we create a new column: 'serial_num' where we assign serial numbers to each user id
+            if print_progress:
+                tqdm.pandas(desc="User location generation")
+                ret_stps = ret_stps.groupby("user_id", as_index=False).progress_apply(
+                    _generate_locations_per_user,
+                    geo_col=geo_col,
+                    distance_metric=distance_metric,
+                    db=db,
+                )
+            else:
+                ret_stps = ret_stps.groupby("user_id", as_index=False).apply(
+                    _generate_locations_per_user,
+                    geo_col=geo_col,
+                    distance_metric=distance_metric,
+                    db=db,
+                )
+
+            # sort so that the last location id of a user = max(location id)
+            ret_stps = ret_stps.sort_values(["user_id", "location_id"])
+
+            # identify start positions of new user_ids
+            start_of_user_id = ret_stps["user_id"] != ret_stps["user_id"].shift(1)
 
-                # enforce unique lables across all users without changing noise labels
-                max_label = np.max(labels)
-                labels[labels != -1] = labels[labels != -1] + location_id_counter
-                if max_label > -1:
-                    location_id_counter = location_id_counter + max_label + 1
+            # calculate the offset (= last location id of the previous user)
+            # multiplication is to mask all positions where no new user starts and addition is to have a +1 when a
+            # new user starts
+            loc_id_offset = ret_stps["location_id"].shift(1) * start_of_user_id + start_of_user_id
+
+            # fill first nan with 0 and create the cumulative sum
+            loc_id_offset = loc_id_offset.fillna(0).cumsum()
```

Comment:
This solution works perfectly for solution without noise label (i.e., setting `num_samples=1`). I think we did not deal with the noise label where location_id == -1. Therefore calculating cumsum will create identical `location_id` when switching users (see image below). We need to mask location_id = -1 records. 

You can reproduce the error by setting num_samples to a bigger value, e.g., num_samples=5.

![image](https://user-images.githubusercontent.com/17105123/121923670-7420c000-cd3b-11eb-9d39-24611101eda9.png)

Reply:
@hong2223 , thanks for looking into this. I am having trouble reproducing this. I think I can exactly reproduce it if I get your test script:
1) can you share the test script and any new input file you used? 
2) can you confirm that you don't have this problem with master?

When I tried to change all `num_sampes` to 5, several of the test functions in `test_staypoints.py` crashes with the mater as well.

Reply:
After the latest fix, tested with this file [geolife_stps_10k](https://github.com/mie-lab/trackintel/files/6669303/geolife_stps_10k.csv) file. 

The fix is same as the comment [above](https://github.com/mie-lab/trackintel/pull/232#discussion_r651086632) by you. 
Also added a new [test](https://github.com/abcnishant007/trackintel/blob/0799e9e4e0a1711e062247e356b1c107f292965e/tests/preprocessing/test_staypoints.py#L140) to handle this: